src/Network/Packet.cs

```diff
@@ -199,6 +199,41 @@ public string ReadUTF8StringSafe()
             return sb.ToString();
         }
 
+        public string ReadUTF8StringSafe(int length)
```

Comment:
I changed this code time ago, but never pushed.
I tried it a bit and looks good. I'm not sure if the length checks are really necessary here.
What do you think about? 

```csharp
        public string ReadUTF8StringSafe()
        {
            _sb.Clear();

            if (Position >= Length) 
                return Empty;

            int index = Position;

            while (index < Length)
            {
                byte b = _data[index++];

                if (b == 0)
                    break;
            }

            string s = Encoding.UTF8.GetString(_data, Position, index - Position);
          
            Seek(index);

            index = 0;

            for (int i = 0; i < s.Length && StringHelper.IsSafeChar(s[i]); i++, index++)
            {

            }

            if (index == s.Length - 1)
                return s;

            for (int i = 0; i < s.Length; i++)
            {
                if (StringHelper.IsSafeChar(s[i]))
                    _sb.Append(s[i]);
            }

            return _sb.ToString();
        }
```

Reply:
The reason I needed a version that specifies length was because book data packets have fixed-length character arrays rather than null-terminated strings. They aren't guaranteed to end in `\0`.

Unrelated to this PR, but UTF-8 is kind of backwards-compatible with ASCII anyway, so perhaps `ReadASCII()` should be done away with for most cases in favour of UTF-8 (but not necessarily `ReadUTF8StringSafe()`, because in some situations the "unsafe" chars could be needed).